### PR TITLE
DAO-171 Fix Timer component to immediately show the correct countdown

### PR DESCRIPTION
**Pull request by mcoetzee**

### CODE REVIEW

File: src/components/timer/timer.tsx [hunk 1]
```diff
@@ -17,36 +17,26 @@ export const DATE_FORMAT = 'do MMMM yyyy';
...
+  const [countdown, setCountdown] = useState<Countdown>(() => calculateCountdown(deadline));
 
   useEffect(() => {
     const timer = setInterval(() => {
-      const now = new Date().getTime();
-      const deadlineDate = deadline.getTime();
+      const countdown = calculateCountdown(deadline);
+      setCountdown(countdown);
 
-      let distance = deadlineDate - now;
```
Comment (Siegrift) [inline src/components/timer/timer.tsx:33]: I am not really sure why I called it `distance` previously.

**Review by Siegrift — APPROVED** [review]
👍 Thanks. Your implementation is much better (and actually works).

File: src/components/timer/timer.tsx [hunk 2]
```diff
@@ -17,36 +17,26 @@ export const DATE_FORMAT = 'do MMMM yyyy';
...
+}
+
 const Timer = (props: Props) => {
   const { deadline, size = 'normal', showDeadline, onDeadlineExceeded } = props;
-  const [timerDays, setTimerDays] = useState('00');
-  const [timerHours, setTimerHours] = useState('00');
-  const [timerMinutes, setTimerMinutes] = useState('00');
-  const [timerSeconds, setTimerSeconds] = useState('00');
-  const [dateDiff, setDateDiff] = useState(0);
+  const [countdown, setCountdown] = useState<Countdown>(() => calculateCountdown(deadline));
```
[the remark this answers is not in the document]
Comment (mcoetzee) [reply src/components/timer/timer.tsx:30]: It's a useful feature of `useState()` when you don't want to calculate the initial state on each render.

E.g. `calculateCountdown()` gets called on every render:
```ts
useState<Countdown>(calculateCountdown(deadline));
```
E.g. `calculateCountdown()` only gets called once when initialising: 
```ts
useState<Countdown>(() => calculateCountdown(deadline));
```

[the remark this answers is not in the document]
Comment (Siegrift) [reply src/components/timer/timer.tsx:30]: Oh, I see.

File: src/components/timer/timer.test.tsx [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import Timer from './timer';
+
+describe('<Timer />', () => {
+  it('shows that it has ended when in the past', () => {
+    const deadline = addSeconds(new Date(), -1);
+
+    render(<Timer deadline={deadline} />);
+
+    expect(screen.getByTestId('timer')).toHaveTextContent('Ended');
+    expect(screen.getByTestId('timer')).toHaveTextContent('00D:00HR:00MIN');
```
[the remark this answers is not in the document]
Comment (mcoetzee) [reply src/components/timer/timer.test.tsx:12]: Hmm regardless of what you use (be it Cypress, React Testing Lib, Playwright etc), you will still need to assert the text content of the DOM. So I don't agree 🙂

We could of course add more fine grained assertions by adding more test ids like:
```ts
expect(screen.getByTestId('timer-days')).toHaveTextContent('01');
expect(screen.getByTestId('timer-hours')).toHaveTextContent('13');
expect(screen.getByTestId('timer-minutes')).toHaveTextContent('30');
```

[the remark this answers is not in the document]
Comment (Siegrift) [reply src/components/timer/timer.test.tsx:12]: >  you will still need to assert the text content of the DOM. So I don't agree 🙂

That's fair. 

> We could of course add more fine grained assertions by adding more test ids like:

I think your implementation is fine as is.

**Pull request merged**